### DISCUSSION TITLE
How about adding the ability to calibrate DC volts to nimi-python?

### DISCUSSION THREAD

**Discussion by dbbotkin** [ni/nimi-python#1806] [Ideas]
I support a charity that refurbishes out-of-date meters and gear for use by kids in educational programs. This involves verification and calibration, particularly after work has been done. The master voltage reference we use is a Fluke 732a. A recent donation of a NI 4070 needs a real calibration (self cal is only good for "tempco" errors). 

How difficult would it be to adapt the code one of current functions to this purpose? I have read the NI calibration guide and there are LabVIEW models and C code for the API. I would be happy to assist and test as you wish.

**Comment by marcoskirsch** [comment]
Adding calibration support to an API in nimi-python is not that straightforward unfortunately. It requires redirecting to a different init function, adding several methods and properties. The hardest part is automating the validation.

It's also an explicit non-goal, as you can see [here](https://github.com/ni/nimi-python/wiki/API-design-guidelines#deviate-from-c-apis-in-the-following-ways).

The 4070 is a difficult device to correctly calibrate. It requires a good deal of equipment, for example in order to adust AC flatness, and a temperature controlled room.

My advice is to work with NI to see if they will help you calibrate the board. Alternatively you can try to get a LabVIEW license or just calibrate with C. But really it might be easiest to ship it to NI.

**Comment by dbbotkin** [comment]
Hello Marcos,
Thanks for the response and the link to the 'statement of objectives, constraints and non-goals’ the last being a fully justified protection. My situation might be different in that my work ‘in retirement’ involves refreshing eBay gear. The idea being to keep it out of landfills and donate it to youth educational activities. I have a temperature controlled room (+/- 1C)  and the part about “a good deal of equipment’ got my wife laughing . . . . 

At the top are a Fluke 732a, an HP3458a for DC measurement. Also an HP3245a as a source of precision low voltage DC and AC. In addition to ethernet, there is a GPIB network that ties into NI MAX so all the gear can do their part as needed. Right now I’m working on a Python script to link the 3245, 3458 and the NI 4070 to verify DC voltage, ultimately along the lines of the NI guide for the device.

According to what I’ve uncovered from NI documents and video, the DC calibration might be accessed via a call to a LabVIEW VI. One of the outstanding features of the HP3458a is that calibration and adjustment of everything in DC can be done with three references: a thermal short, a 10 V source and a 10k ohm resistor. The DC cal of the 4070 has a couple more stages, also doable. Resistance cal requires more standards, but those are also available. I don’t have a calibrator because affordable ones are not to be found in good condition where I shop. As my task is not urgent, I can do it 'old-school' with transfer standards. 

I am not a metrologist, however I have made it a practice to learn as much as I can to be aware of the sources of error. Ultimately, sending primary standards in to a real Cal Lab is on the plan, so making the call to NI is a good suggestion.

Finally, I am new to Github and have forked the Nimi-Python repository. As I develop python scripts, should these be posted on my fork or what?

Cheers,
Donal

**Reply by marcoskirsch** [reply]
> the part about “a good deal of equipment’ got my wife laughing . . . .

That is awesome 🤣

> Finally, I am new to Github and have forked the Nimi-Python repository. As I develop python scripts, should these be posted on my fork or what?

I do suggest that you put scripts that you believe are useful to yourself and others on GitHub. We keep our examples lean on `nimi-python` – we're trying to get people started with the APIs, not teach them how to use the instruments comprehensively. Maybe a new, separate repo is the way to go.

If you do believe that there's a big need for an example in `nimi-python` or you make an improvement that should go into the main repo, then make your change on your fork and do a PR. It's great if you can also open issues on the main repo and/or ask questions here. Even if no action comes out of them, the fact that they are documented is helpful.

Thanks!
